### PR TITLE
[MM-55993] Create a composable independent table to be used in Admin reporting

### PR DESCRIPTION
**Pull request by M-ZubairAhmed**
#### Summary
- Added a new library "@tanstack/react-table"
- Created a table component to be used in system's console

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55993

#### Screenshots
Screenshot 2023-12-17 at 10 11 29

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: webapp/channels/src/components/admin_console/system_users/sample.ts
Comment (M-ZubairAhmed) [inline webapp/channels/src/components/admin_console/system_users/sample.ts:1]: Will be removed before merge

File: webapp/channels/src/sass/utils/_mixins.scss
Comment (M-ZubairAhmed) [inline webapp/channels/src/sass/utils/_mixins.scss:77]: star property hack used for ie7 and below

Comment (devinbinnie) [reply webapp/channels/src/sass/utils/_mixins.scss:77]: Are we supporting IE7 with this? Can we not just remove it?

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Reviewing this while @abhijit is out. It's looking great - just a few minor things I've noticed:

1. The header for the column "Messages posts" should be "Messages posted"

2. For the vertical borders on the user details column and actions column, is it possible to hide those borders when they aren't needed and only show them when content is getting cut off due to screen width?

3. The content in the user details column doesn't align with the header for the column.
    Screenshot 2023-12-12 at 6 16 31 PM

4. Content in the Email column feels too tight to the left edge. Can we adding some more breathing room there? I'd say at least 4px more on the left of all cells would help.
    Screenshot 2023-12-12 at 6 18 01 PM

5. The header for "Actions" is not aligned with the text in the buttons below it. Can we adjust this? I realize this is because of the padding in each of the buttons. But we can just adjust the padding on the left of the Actions header and force them to align. We could also consider overriding the default padding on these buttons to reduce it ti 12px instead of 16px for this context.
    Screenshot 2023-12-12 at 6 21 57 PM

6. As mentioned in the thread, I was wondering about the pagination. Should we consider having it at the top and the bottom of the table? Depending on the number of rows per page, you have to scroll down pretty far to access the pagination. Since we can't show 'X of Y users' at the top anyway, maybe we can place the pagination there as well as at the bottom.

7. In the pagination, the < and > buttons are spaced much further apart from what is designed. The buttons also seem much wider. I think these should be using the `btn btn-icon btn-sm` classes rather than `btn btn-sm`

image

File: webapp/channels/src/sass/utils/_mixins.scss
Comment (M-ZubairAhmed) [reply webapp/channels/src/sass/utils/_mixins.scss:77]: we were supporting ie7 with that, thats why i removed it since we are no longer supporting any version of ie

File: webapp/channels/src/components/admin_console/admin_console_list_table/index.tsx
Comment (harshilsharma63) [inline webapp/channels/src/components/admin_console/admin_console_list_table/index.tsx:1]: The copyrights added.

**Review by harshilsharma63 — APPROVED** [review]

**Comment by M-ZubairAhmed** [comment]
@matthewbirtch Apart from
> For the vertical borders on the user details column and actions column, is it possible to hide those borders when they aren't needed and only show them when content is getting cut off due to screen width?

All the concerns are corrected. This one I guess could be done in a later stage because it's a bit complex to work on.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
A couple other change requests:

1. The table extends beyond the main container when the browser width is smaller
    - also, when I scroll horizontally in this scenario, the header and footer scroll with it and don't remain in a fixed position. I don't recall seeing this in the last review though.

    image

    https://github.com/mattermost/mattermost/assets/2040554/ffe8c0ae-a291-43a5-97dd-0be1b17ddecc

3. Can we make the number that appears in the select input for show X users per page fully opaque center channel color? It sort of feels like it is disabled right now. Same with the options when you open the menu. 
    image

File: webapp/channels/src/sass/utils/_mixins.scss
Comment (devinbinnie) [reply webapp/channels/src/sass/utils/_mixins.scss:77]: Ahh the remnants always linger :P

File: webapp/channels/src/components/admin_console/list_table/elapsed_duration_cell.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/list_table/elapsed_duration_cell.tsx:38]: Nit: can we move this to the top? No use instantiating anything if we're not going to render anything anyways.

File: webapp/channels/src/components/admin_console/system_users/system_users_cell_elapsed_days/index.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/system_users/system_users_cell_elapsed_days/index.tsx:38]: Same here.

File: webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx:80]: Do we need a blank string here? If `display_name` is `null` or `undefined` it would just show nothing right?

File: webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx:212]: I assume we're removing these?

File: webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx:26]: This is more of a question for my own knowledge: is there a reason to prefer `formatMessage` the function vs `FormattedMessage` the component? I guess I can see the function if we have some data that we need to pass in that changes based on a loop or some other non-props related factor, but I would assume the component is better in most cases.

0/5 on changing it though, I was just curious.

File: webapp/channels/src/components/admin_console/system_users/system_users.tsx
Comment (devinbinnie) [inline webapp/channels/src/components/admin_console/system_users/system_users.tsx:336]: I assume this will be hooked up eventually, are we waiting to do the filter component re-design in a follow-up PR?

**Review by devinbinnie — CHANGES_REQUESTED** [review]
Looking good, just a few comments/questions

File: webapp/channels/src/components/admin_console/system_users/system_users.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/system_users/system_users.tsx:336]: yes this is just more of a design review pr for the table. filters are specific to the implementation instead of being part of generic table.

File: webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx:212]: yes this file will not be merged

File: webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx:80]: Yes library expects the data if to be present if not them empty string.

File: webapp/channels/src/components/admin_console/list_table/elapsed_duration_cell.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/list_table/elapsed_duration_cell.tsx:38]: we cant move this to the top because we have hooks defined here, conditional rendering of hooks is a no-no.

File: webapp/channels/src/components/admin_console/system_users/system_users_cell_elapsed_days/index.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/system_users/system_users_cell_elapsed_days/index.tsx:38]: file is no more used

File: webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx
Comment (M-ZubairAhmed) [reply webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx:26]: I tend to use function over components in places where the addition of a span tag is of no concern (which is created with the use of component). So yes in this case we could just use the component

File: webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/admin_console/system_users/system_users_list/system_users_list.tsx:212]: Okay ping me once this is removed :)

File: webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx
Comment (devinbinnie) [reply webapp/channels/src/components/admin_console/system_users/system_users_list_actions/index.tsx:26]: Okay, I'll leave it to you to make the decision on what to use. Thanks for the insight.

**Review by matthewbirtch — APPROVED** [review]
Looks like the changes I noted are resolved. Thanks @M-ZubairAhmed. I'll let @abhijit-singh finish off the UX review from here

**Comment by M-ZubairAhmed** [comment]
@abhijit-singh gentle ping on this one

**Review by abhijit-singh — APPROVED** [review]
Looks great @M-ZubairAhmed! A few requests but mostly minor and non-blocking:
1. Can we make the spacing on all sides of the table consistent to 20px? Currently it seems that on the right, it is just 10 px. Changing the margin-right here from 10px to 20px should work. 
image
2. If there's just 1 page of users, can we completely disable the next/previous icon buttons so that they're not clickable at all — No hover state, no on-click state. 
image
3. Can we add a slight box-shadow to the overall table container(`admin-console__container`) as in the design?
Here's what I had added in the design: 
```
/* Elevation 1 */
box-shadow: 0px 2px 3px 0px rgba(0, 0, 0, 0.08);
```


Also, I'm assuming there'll be other PRs to make the sorting work, update filters and add the total user count. Looking forward to reviewing those.

**Pull request merged**
